Change local storage and synchronization mechanisms

sedationh opened the pull request:
With the addition of image support, `page_elements` can now be quite large (over 5 MB). Therefore, `localStorage` has been replaced with `IndexedDB`, and synchronization is now active/manual. This addresses issues with the previous method being very slow and conflict-prone.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added Google OAuth login option to both login and sign-up forms.
  * Added a manual download button for users to fetch the latest data from the server, overwriting unsaved changes.

* **Refactor**
  * Switched data storage to use asynchronous IndexedDB-backed storage, improving data persistence.
  * Changed data loading and saving to be explicitly user-triggered with debounced autosave, replacing automatic periodic saves.
  * Optimized data queries to fetch only necessary page details, enhancing performance.
  * Improved route tree generation with clearer naming conventions and reorganized typings.

* **Bug Fixes**
  * Improved error handling and user feedback when fetching data from the server.

* **Chores**
  * Updated dependencies to include ahooks, dexie, and localforage for enhanced storage and utility capabilities.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

sedationh committed:
feat: enhance drawing functionality with file support

- Added support for BinaryFiles in ExcalidrawData type.
- Updated createNewPage and setDrawData functions to handle files.
- Modified drawDataStore to include files in state management.
- Enhanced Mermaid and Page components to manage file uploads and updates.

sedationh committed:
Merge branch 'macintushar:main' into main

sedationh committed:
feat: integrate localforage for persistent storage in drawDataStore

- Added localforage to manage persistent storage for drawing data.
- Updated drawDataStore to utilize localforage for storing and retrieving page data.
- Enhanced Page component to fetch data from the server and update local storage accordingly.
- Modified UI to include a button for pulling the latest data from the server.

sedationh committed:
optimize: return only necessary fields in getPages query

- Updated getPages function to select specific fields (page_id, name, created_at, updated_at)
- Avoid returning large page_elements data to improve performance
- Reduces bandwidth usage when fetching page list

sedationh committed:
feat: add Google OAuth login functionality

- Implemented signInWithGoogle function in auth module for Google authentication.
- Updated Login and SignUp components to include Google login button and handle Google login process.
- Added loading state management for Google login in both components.
- Enhanced user experience with toast notifications for login errors.

sedationh committed:
refactor: update comment for clarity in Page component

- Changed comment from Chinese to English for better understanding.
- Ensured consistency in code documentation.

sedationh committed:
feat: implement auto-save functionality in Page component

- Added useRef to manage save timeout for scene changes.
- Implemented handleSceneChange to debounce and save drawing data to local storage automatically.
- Included cleanup effect to clear timeout on component unmount.

sedationh committed:
refactor: simplify saving logic in Page component

- Removed redundant checks before saving drawing data.
- Streamlined the saving process to always save locally first, followed by pushing to the API.
- Improved code readability by adjusting formatting in the setPageData call.

macintushar commented on src/db/auth.ts:
```diff
@@ -8,6 +8,16 @@ export async function login(email: string, password: string) {
   return { data, error };
 }
 
+export async function signInWithGoogle() {
```
@sedationh Google OAuth is not in the roadmap currently. There's more implications to this such as with the metadata that need to be tested before doing this.

sedationh replied:
got it, but I truly need this feature, if you do think so, i can checkout another branch to pull request

sedationh replied:
you could expriece it in my deployment. https://draw.sedationh.top/

I do not find problem until now

macintushar replied:
@sedationh can you separate out the code for Google OAuth and raise a separate PR for it?

macintushar commented:
@sedationh data is not loading when you click on a Page.

sedationh commented:
> @sedationh data is not loading when you click on a Page.

yeah, that is a problem, when local has no files, download remote files, I will fix it tomorrow

sedationh committed:
feat: add ahooks library and enhance Page component with loading state

- Added "ahooks" library for improved async handling.
- Introduced loading state management in the Page component to indicate data fetching.
- Updated scene loading logic to provide user feedback during data retrieval.

sedationh committed:
refactor: move cleanup effect for save timeout in Page component

- Moved the cleanup effect for clearing the save timeout to a dedicated useEffect.
- Removed redundant cleanup code that was previously commented out.
- Improved code organization and readability in the Page component.

sedationh committed:
refactor: simplify loading overlay logic in Page component

- Consolidated loading overlay condition to improve readability.
- Enhanced the structure of the loading state rendering in the Page component.

sedationh committed:
refactor: enhance success and error handling in Page component

- Updated success callback to display a toast notification upon successful data save.
- Consolidated the logic for setting the saving state to ensure it is reset after the mutation settles.
- Improved button loading state management for better user feedback during save operations.

sedationh committed:
feat: integrate Dexie for improved data management in drawDataStore

- Replaced localforage with Dexie for persistent storage in drawDataStore.
- Updated setPageData and getPageData methods to work with Dexie for better performance and reliability.
- Refactored Page component to handle asynchronous data operations with the new store structure.

macintushar commented on src/views/Page.tsx:
```diff
@@ -40,158 +47,212 @@ export default function Page({ id }: PageProps) {
+                      onClick={fetchDataFromServer}
+                      disabled={isFetching}
+                    >
+                      <Download className="h-4 w-4" />
```
@sedationh let's change this back to the original RefreshCcw. Download implies that the Excalidraw File or Image will be downloaded.

macintushar commented on src/views/Page.tsx:
Seems like the Autosave Logic is removed? Why was that done?

macintushar commented on pnpm-lock.yaml:
Please remove this pnpm-lock file

macintushar commented on package.json:
I don't see localforage and ahooks used anywhere in the code, that can be removed.
